engine/execution/ingestion/stop/stop_control.go

```diff
@@ -0,0 +1,660 @@
+
+// validateStopChange verifies if the stop parameters can be changed
+// returns the error with the reason if the parameters cannot be changed.
+//
+// Stop parameters cannot be changed if:
+//   - node is already stopped
+//   - stop parameters are immutable (due to them already affecting execution see
+//     ShouldExecuteBlock)
+//   - stop parameters are already set by a different source and the new stop is later than
+//     the existing one
+//
+// Expected error returns during normal operations:
+//   - ErrCannotChangeStop: this indicates that new stop parameters cannot be set.
+//
+// Caller must acquire the lock.
+func (s *StopControl) validateStopChange(
+	newStopBoundary stopBoundary,
+) error {
+
+	errf := func(reason string) error {
+		return fmt.Errorf("%s: %w", reason, ErrCannotChangeStop)
+	}
+
+	if s.stopped {
+		return errf("cannot update stop parameters, already stopped")
+	}
+
+	if s.stopBoundary.immutable {
+		return errf(
+			fmt.Sprintf(
+				"cannot update stopHeight, stopping commenced for %s",
+				s.stopBoundary),
+		)
+	}
+
+	if !s.stopBoundary.Set() {
+		// if the current stop is no stop, we can always update
+		return nil
+	}
+
+	if s.stopBoundary.fromVersionBeacon == newStopBoundary.fromVersionBeacon {
+		// if the stop was set by the same source, we can always update
+		return nil
+	}
+
+	// if one stop was set by the version beacon and the other one was manual
+	// we can only update if the new stop is strictly earlier
+	if newStopBoundary.StopBeforeHeight < s.stopBoundary.StopBeforeHeight {
+		return nil
+
+	}
+	// this prevents users moving the stopHeight forward when a version newStopBoundary
+	// is earlier, and prevents version beacons from moving the stopHeight forward
+	// when a manual stop is earlier.
+	return errf("cannot update stopHeight, " +
+		"new stop height is later than the current one")
+}
+
+// GetStopParameters returns the upcoming stop parameters or nil if no stop is set.
+func (s *StopControl) GetStopParameters() StopParameters {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.stopBoundary.StopParameters
+}
+
+// ShouldExecuteBlock should be called when new block can be executed.
+// The block should not be executed if its height is above or equal to
+// s.stopBoundary.StopBeforeHeight.
+//
+// It returns a boolean indicating if the block should be executed.
+func (s *StopControl) ShouldExecuteBlock(b *flow.Header) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	// don't process anymore blocks if stopped
+	if s.stopped {
+		return false
+	}
+
+	// Skips blocks at or above requested stopHeight
+	// doing so means we have started the stopping process
+	if b.Height < s.stopBoundary.StopBeforeHeight {
+		return true
+	}
+
+	s.log.Info().
+		Msgf("Skipping execution of %s at height %d"+
+			" because stop has been requested %s",
+			b.ID(),
+			b.Height,
+			s.stopBoundary)
+
+	// stopBoundary is now immutable, because it started affecting execution
+	s.stopBoundary.immutable = true
```

Comment:
This is very counter intuitive. Why do we set `immutable` but not `stopped`? We still return `false` for should we  process the block meaning we have reached a threshold but don't set `stopped` to true. I assume it's because of finalization? If there is some specific logic for it we better explain this.

Reply:
added comment.